test/object/object.js

```diff
@@ -157,4 +157,22 @@ function test(binding) {
     assert.strictEqual(binding.object.instanceOf({}, Ctor), false);
     assert.strictEqual(binding.object.instanceOf(null, Ctor), false);
   }
+
+  {
+    const obj = {
```

Comment:
The test case here didn't reflect the exceptional path of JavaScript. Maybe we can test with proxies with throwing handlers:

```javascript
const o = new Proxy({}, {
  getOwnPropertyDescriptor(target, p) {
    // throwing primitives is valid in javascript;
    throw 'eh-oh';
  },
  ownKeys(target) {
    throw '¯\_(ツ)_/¯';
  },
});
```

Reply:
@legendecas Running this locally produces this log:
```console
❯ node --trace-uncaught test/object/object.js

/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:175
        throw '¯\_(ツ)_/¯';
        ^
¯_(ツ)_/¯
Thrown at:
    at ownKeys (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:175:9)
    at test (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:180:23)
    at /home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:5:1
    at Module._compile (node:internal/modules/cjs/loader:1102:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1131:10)
    at Module.load (node:internal/modules/cjs/loader:967:32)
    at Module._load (node:internal/modules/cjs/loader:807:14)
    at executeUserEntryPoint (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
❯ echo $?
1
```
Should I write a test that asserts that this throws `/¯_\(ツ\)_\/¯/`?

Reply:
The failure you encountered is probably the for-loop in JavaScript. What I mean is getting the proxied object to the native looping to see if the native side is handling the exception in a reasonable way.

Reply:
@legendecas That is correct. I tried to call the function directly and it crashes because the iterator internally tries to get a copy of the keys of the object using `GetPropertyNames` and that throws the exception:

<details>

```console
❯ node test/object/object.js
FATAL ERROR: Error::Error napi_create_reference
 1: 0xacd720 node::Abort() [node]
 2: 0xaced2e node::OnFatalError(char const*, char const*) [node]
 3: 0xacedc1  [node]
 4: 0xa954f4 napi_fatal_error [node]
 5: 0xb5048821  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 6: 0xb5048b0d Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 7: 0xb508f5af GetPropertyNames(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 8: 0xb5096756 Napi::details::CallbackData<Napi::Array (*)(Napi::CallbackInfo const&), Napi::Array>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
...
```

</details>

However, this code too crashes and produces a similar trace if we add this in the same file (`test/object/object.js`):

```js
  {
    const o = new Proxy({}, {
      getOwnPropertyDescriptor(target, p) {
        // throwing primitives is valid in javascript;
        throw 'eh-oh';
      },
      ownKeys(target) {
        throw '¯\_(ツ)_/¯';
      },
    });
    binding.object.GetPropertyNames(o);
  }
```

<details>

```console
❯ node test/object/object.js
FATAL ERROR: Error::Error napi_create_reference
 1: 0xb37720 node::Abort() [node]
 2: 0xb38d2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb38dc1  [node]
 4: 0xaff4f4 napi_fatal_error [node]
 5: 0xb5048821  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 6: 0xb5048b0d Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 7: 0xb508f5af GetPropertyNames(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
 8: 0xb5096756 Napi::details::CallbackData<Napi::Array (*)(Napi::CallbackInfo const&), Napi::Array>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding.node]
...
```

</details>

Is this the expected behaviour?

Reply:
Oh, my bad. Please note that it's the primitives that cause the crash. That is a known issue. The case you may need to take into consideration is _thwoing_ during the iteration:

```javascrpt
const o = new Proxy({}, {
  getOwnPropertyDescriptor(target, p) {
    throw new Error();
  },
  ownKeys(target) {
    throw new Error();
  },
});
```

Reply:
@legendecas Thanks for updating the example. I tried to run this on `binding.object.sum()` and it produces the expected error:

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    binding.object.sum(obj);
  }
```

```console
❯ node test/object/object.js
/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:178
    binding.object.sum(obj);
                   ^

Error: ownKeys error
    at Object.ownKeys (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:174:15)
    at test (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:178:20)
    at Object.<anonymous> (/home/raisinten/Desktop/universe/git/node-addon-api/test/object/object.js:5:1)
    at Module._compile (node:internal/modules/cjs/loader:1102:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1131:10)
    at Module.load (node:internal/modules/cjs/loader:967:32)
    at Function.Module._load (node:internal/modules/cjs/loader:807:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
```

</details>

But when I try to assert that it throws the particular error, the process seems to crash, which does not happen when I try to assert that `binding.object.GetPropertyNames(obj);` too throws the error.

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    assert.throws(() => {
      binding.object.sum(obj);
    }, /ownKeys error/);
  }
```

```console
❯ node test/object/object.js
FATAL ERROR: Error::New napi_get_last_error_info
 1: 0xb11720 node::Abort() [node]
 2: 0xb12d2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb12dc1  [node]
 4: 0xad94f4 napi_fatal_error [node]
 5: 0xb4f5e841  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 6: 0xb4f68a7a Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 7: 0xb504589d Sum(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 8: 0xb4f68c70 Napi::details::CallbackData<Napi::Value (*)(Napi::CallbackInfo const&), Napi::Value>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
```

</details>

Since the current code does not attempt to catch any exception, shouldn't the exception propagate to the JS level anyway? I don't understand why there is a change in behaviour of the error when wrapping the thrower with an `assert.throws`.

Reply:
The JavaScript exception is thrown at the `Object::iterator::iterator`. When c++ exceptions are disabled, the native code will continue to run if there are JavaScript exceptions thrown. And when JavaScript exception is pending, node-api would complain on most API calls, i.e. returning `napi_exceptions_pending`. So when `Object::iterator::iterator` triggered exceptions, the following steps like `Object::const_iterator::operator*` would abort subsequent `napi_exceptions_pending`.

Reply:
@legendecas I tried to explicitly catch, clear and throw any pending exceptions in the dereferencing operator overloads but still I'm getting the same error:

<details>

```diff
diff --git a/napi-inl.h b/napi-inl.h
index c9f3ab1..54d7b52 100644
--- a/napi-inl.h
+++ b/napi-inl.h
@@ -1423,6 +1423,11 @@ inline std::pair<Value, Object::PropertyLValue<Value>>
 Object::const_iterator::operator*() const {
   Value key = _keys[_index];
   const PropertyLValue<Value> value = (*_object)[key];
+  Napi::Env env(_object->_env);
+  if (env.IsExceptionPending()) {
+    Error e = env.GetAndClearPendingException();
+    NAPI_THROW(e, {key, value});
+  }
   return {key, value};
 }
 
@@ -1459,6 +1464,11 @@ inline std::pair<Value, Object::PropertyLValue<Value>>
 Object::iterator::operator*() {
   Value key = _keys[_index];
   const PropertyLValue<Value> value = (*_object)[key];
+  Napi::Env env(_object->_env);
+  if (env.IsExceptionPending()) {
+    Error e = env.GetAndClearPendingException();
+    NAPI_THROW(e, {key, value});
+  }
   return {key, value};
 }
 
```

```console
❯ node test/object/object.js
FATAL ERROR: Error::New napi_get_last_error_info
 1: 0xb4b720 node::Abort() [node]
 2: 0xb4cd2e node::OnFatalError(char const*, char const*) [node]
 3: 0xb4cdc1  [node]
 4: 0xb134f4 napi_fatal_error [node]
 5: 0xb4e94871  [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 6: 0xb4e9eaaa Napi::Error::New(napi_env__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 7: 0xb4f7ba85 Sum(Napi::CallbackInfo const&) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
 8: 0xb4e9eca0 Napi::details::CallbackData<Napi::Value (*)(Napi::CallbackInfo const&), Napi::Value>::Wrapper(napi_env__*, napi_callback_info__*) [/home/raisinten/Desktop/universe/git/node-addon-api/test/build/Release/binding_noexcept.node]
...
```

</details>

What am I doing wrong?

Reply:
The key is that:

> When c++ exceptions are disabled, the native code will continue to run even if there are JavaScript exceptions thrown.

Reply:
@legendecas But then, why does this run without crashing?

<details>

```js
  {
    const obj = new Proxy({
      '-forbid': -0x4B1D,
      '-feedcode': -0xFEEDC0DE,
      '+office': +0x0FF1CE,
      '+forbid': +0x4B1D,
      '+deadbeef': +0xDEADBEEF,
      '+feedcode': +0xFEEDC0DE,
    }, {
      getOwnPropertyDescriptor(target, p) {
        throw new Error("getOwnPropertyDescriptor error");
      },
      ownKeys(target) {
        throw new Error("ownKeys error");
      },
    });

    assert.throws(() => {
      binding.object.GetPropertyNames(obj);
    }, /ownKeys error/);
  }
```

</details>

`GetPropertyNames` too is being run with C++ exceptions disabled, so shouldn't the behaviour be the same?

Reply:
The key is on "the native code will continue to run regardless of pending JavaScript exceptions". When there is only a single "continuous" node-api call, that's fine. When returning to node-core/v8 with pending JavaScript, the exception is routed to JavaScript. However, when there are multiple sequential continuous native node-api that calling into JavaScript and when the c++ exception is disabled, there is no way to break the native code execution on pending JavaScript, and those calls after the one caused the JavaScript exception would abort since they are unable to handle the situation - we can not throw again while there is a pending JavaScript exception.

Back to the case here, we have multiple continuous node-api calls in the iterators:
- one for getting object names (i.e., proxy handler `ownKeys`, https://github.com/nodejs/node-addon-api/pull/930/files#diff-cd0b229bcaf532206661e2ee1a603a7f60ac22ecfcb239920f6b42a34d909f34R1393) 
- and one for getting the first value of the keys array (https://github.com/nodejs/node-addon-api/pull/930/files#diff-cd0b229bcaf532206661e2ee1a603a7f60ac22ecfcb239920f6b42a34d909f34R1424)

Those two procedures are both located in node-addon-api so there is no way for user code to break on exceptions. There comes the crash.

Reply:
@legendecas Thank you for the awesome explanation. I will enclose the iterator code in `#ifdef NAPI_CPP_EXCEPTIONS` guards. How do I locally build and test this with C++ exceptions enabled?

Reply:
By default `npm test` will run the tests on both flavors. See https://github.com/nodejs/node-addon-api/blob/main/test/binding.gyp for build configurations and https://github.com/nodejs/node-addon-api/blob/main/test/common/index.js#L78 for test runners.

Reply:
@legendecas Hi! I have made the changes so that the iterator is available only when we are building with c++ exceptions enabled and it passes the tests locally. PTAL.

Reply:
ehhh... I was using the older test configuration while testing locally, that's why it is failing to require the module on GH Actions. Any neat way to tell the JS test whether the c++ exceptions is enabled?

Reply:
> Any neat way to tell the JS test whether the c++ exceptions is enabled?

In `test/common/index.js` there is a function named `runTest` that enumerated all available build flavors to run tests. You can create a subset or options to filter out the c++ exceptions enabled flavors to run the new cases.

Reply:
@legendecas Thank you. I've updated my change. CI is green now.